apps/smart-contracts/core/test/snapshots.ts

```diff
@@ -0,0 +1,79 @@
+/* eslint max-classes-per-file: 0 */
+import { ethers, network } from 'hardhat'
+
+class Snapshot {
+  public snapshotId: string
+
+  constructor(public name: string, public debug: boolean = false) {}
+
+  public async snapshot(): Promise<void> {
+    this.snapshotId = await ethers.provider.send('evm_snapshot', [])
+    if (this.debug) console.log(`💾 Snapshot taken - ${this.name} @ ${this.snapshotId}`)
+  }
+
+  public async reset(): Promise<void> {
+    if (this.debug) console.log(`⚓ Reset snapshot - ${this.name} @ ${this.snapshotId}`)
+    await network.provider.send('evm_revert', [this.snapshotId])
+    this.snapshotId = await ethers.provider.send('evm_snapshot', [])
+  }
+}
+
+export class Snapshotter {
+  // Snapshot stack and setup code. This allows us to nest snapshotting blocks
+  // and have them all work together.
+  private outerSetupComplete = false
+  private snapshotStack: Snapshot[] = []
+  private setActiveSnapshot(snapshot: Snapshot): void {
+    this.snapshotStack.push(snapshot)
+  }
+  private headSnapshot(): Snapshot {
+    return this.snapshotStack[this.snapshotStack.length - 1]
+  }
+  private async popActiveSnapshot(): Promise<void> {
+    // Pop the snapshot
+    this.snapshotStack.pop()
+    if (this.snapshotStack.length > 0) {
+      // If we have another in the stack, revert the blockchain to this snapshot
+      await this.headSnapshot().reset()
+    } else if (this.snapshotStack.length === 0) {
+      // Otherwise we've popped the last one off the stack, so we're done, and our
+      // outer setup is no longer complete.
+      this.outerSetupComplete = false
+    }
+  }
+
+  // This is the actual function which sets up snapshotting for a given block.
+  public usesCustomSnapshot(name: string | undefined = undefined): void {
```

Comment:
```suggestion
  public setupSnapshotContext(name: string | undefined = undefined): void {
```
This still hasn't been changed